Sources/SwiftMemcache/MemcachedConnection.swift

```diff
@@ -0,0 +1,113 @@
+///
+/// This actor can be used to send commands to the server.
+public actor MemcachedConnection {
+    private typealias StreamElement = (MemcachedRequest, CheckedContinuation<MemcachedResponse, Error>)
+    /// The underlying channel to communicate with the server.
+    private let channel: NIOAsyncChannel<ByteBuffer, ByteBuffer>
+    /// The allocator used to create new buffers.
+    private let bufferAllocator: ByteBufferAllocator
+    /// The channel's event loop group.
+    private let eventLoopGroup: EventLoopGroup
+    /// The stream of requests to be sent to the server.
+    private let requestStream: AsyncStream<StreamElement>
+    /// The continuation for the request stream.
+    private let requestContinuation: AsyncStream<StreamElement>.Continuation
+
+    /// Initialize a new MemcachedConnection.
+    ///
+    /// - Parameters:
+    ///   - host: The host address of the Memcache server.
+    ///   - port: The port number of the Memcache server.
+    ///   - eventLoopGroup: The event loop group to use for this connection.
+    public init(host: String, port: Int, eventLoopGroup: EventLoopGroup) async throws {
+        self.eventLoopGroup = eventLoopGroup
+        let bootstrap = ClientBootstrap(group: self.eventLoopGroup)
+            .channelInitializer { channel in
+                return channel.pipeline.addHandlers([MessageToByteHandler(MemcachedRequestEncoder()), ByteToMessageHandler(MemcachedResponseDecoder())])
+            }
+
+        let rawChannel = try await bootstrap.connect(host: host, port: port).get()
+        self.channel = try NIOAsyncChannel<ByteBuffer, ByteBuffer>(synchronouslyWrapping: rawChannel)
+        self.bufferAllocator = ByteBufferAllocator()
+
+        let (stream, continuation) = AsyncStream<StreamElement>.makeStream()
+        self.requestStream = stream
+        self.requestContinuation = continuation
+    }
+
+    /// Start consuming the requestStream
+    ///
+    /// This function starts consuming the requests from the `requestStream`,
+    /// sending each request to the Memcache server and handling the server's responses.
+    public func run() async {
+        for await (request, continuation) in self.requestStream {
+            var buffer = self.bufferAllocator.buffer(capacity: 0)
+            let encoder = MemcachedRequestEncoder()
+            do {
+                try encoder.encode(data: request, out: &buffer)
+                try await self.channel.outboundWriter.write(buffer)
+
+                let responseBuffer = try await self.channel.inboundStream.first { _ in true }
+
+                if var responseBuffer, responseBuffer.readableBytes != 0 {
+                    var decoder = MemcachedResponseDecoder()
+                    let response = try decoder.decode(buffer: &responseBuffer)
+                    continuation.resume(returning: response!)
+                }
+            } catch {
+                continuation.resume(throwing: error)
```

Comment:
If we get an error here it means that our connection is done and we cannot send more requests. We should introduce some state to this actor which keeps track of this and fails any requests.
I recommend introducing this state enum to this actor

```swift
enum State {
    case running(
        /// The allocator used to create new buffers.
        bufferAllocator: ByteBufferAllocator,
        /// The stream of requests to be sent to the server.
        requestStream: AsyncStream<StreamElement>,
        /// The continuation for the request stream.
        requestContinuation: AsyncStream<StreamElement>.Continuation
    )
    case finished
}
```

Then store this state inside the actor as a var. Once we get an error here we are going transition to `finished` and call `requestContinuation.finish()` Whenever we handle a request in this method we have to switch over the state and check that we are not in finished.